Expense Tracker (Workshop Resubmission) (#1651)
hackclub/hackclub: workshops/expense-tracker/README.md

---
name: 'Expense Tracker'
description: 'A command line based expense tracker using Python'
author: '@iamsid47'
img: https://cloud-4tlcr594m.vercel.app/0expen_e_tracker.png
---

## Expense Tracker Using Python

*Hi everyone! In this workshop, we will walk through how to build ***an expense tracker!**** 

The completed project can be found by clicking [here](https://repl.it/@iamsid47/expense-tracker-main).

## End result

![End result](https://cloud-nrw7488qg.vercel.app/0screenshot_2020-12-29_215630.png)

## Let's Get Started

> So, the first thing we need to do is head over to [Repl.it](https://repl.it) and create a *repl*. Choose the **Python** as the language and name your project. I'm going to name mine as **expense-tracker**.

Reviewer: ```suggestion
So, the first thing we need to do is to head over to [Repl.it](https://repl.it) and create a *repl*. Choose **Python** as the language and name your project. I'm going to name mine as **expense-tracker**.
```

![Project Creation](https://cloud-giasrdstj.vercel.app/4expense-tracker.png)

> We will use Repl.it's package managers to get our libraries. For this, head over to the *Packages* and search **matplotlib**. Next, click on the add button to add that package to your repl. Do the same for the **datetime** package. We are also required to have **json** library but it's pre-installed so need to get it again.

Reviewer: ```suggestion
We will use Repl.it's package managers to get our libraries. For this, head over to the *Packages* section and search **matplotlib**. Next, click on the add button to add that package to your repl. Do the same for the **datetime** package. We also need the **json** library but it's pre-installed!
```

> Now, our data is gonna be stored in `.json` format. Thus, let's create a file named: `expenses.json`. In this file, type in a the list named as **Expenses**. This list will keep on updating as we will be adding entries using our *expense-tracker*. Let's open `main.py` and type some code.

Reviewer: add the code for json file as how it should look like

We need to first **import** all the libraries. Thus:

```python
import json
from datetime import datetime
import matplotlib.pyplot as plt
```

> Next, we'll define a function named `write_json` and call the `expenses.json` file.

Reviewer: call? phrasing error

```python
def write_json(data, filename='expenses.json'): 
    with open(filename,'w') as f: 
        json.dump(data, f, indent=4) 
```

Now when the user gives in some input, we need to capture it and store this into the `expenses.json` file which we just created. Also, there should be something displayed when we run our program. So let's print something as well!

```python
expdic = {"Expenses":[]}
dic = {}
print("Howdy! Exp-Track here ;)")

> ```

Reviewer: where should this code come?

After this, we will start taking in the user's inputs. So first when the program runs, we will offer our user some choices. For example, to add in a new purchase or to view past purchase, or just to exit the program.

> Let's add some `choice` then. And later on, some *conditionals* as well. Also, we are gonna accept capital input. Thus, we will add `upper` at the end get the inputs in uppercase.

Reviewer: phrasing mistakes

```python
choice = input("To input a purchase, press Y, To view all Expenses, Press V, else press E to exit: ").upper()
```

> Now here, we add some `if` statements aka conditionals. So, when the user presses **V** we show him the expenses by calling the `expenses.json` file and showing it's content. We also want to generate a graph of our total expenditure, and that's the reason to install the **matplotlib** library. 

Reviewer: "user presses V" this statement is confusing

                        exp = exp + value
                        p.append(value)
            plt.pie(p, labels=p, wedgeprops={'edgecolor': 'black'})
            plt.title("Total Expenditure in Rs.")
            plt.legend(title="Your Expenses",labels=prod, loc='upper center', bbox_to_anchor=(0.5, 0.05),fancybox=True, shadow=True, ncol=5)
            plt.show()
```

To get the total for out expenditure, we create a conditional and make it as *choice* **T**. Inside this, we call the json data which will be store and load it onto a variable named *data*. We again create a variables named **temp** and call just the **Expenses** from the **data** variable which currently contains all the data from the json file. Next, we create a formula to calculate the total expenses.

```python
if choice == "T":
        Expenditure = 0
        with open("expenses.json") as f:
            data = json.load(f)
            temp = data["Expenses"]
            for i in data["Expenses"]:
                for key,value in i.items():
                    if (key == "Price"):
                        Expenditure = Expenditure + value
            print("Your Total Expenditure is ",Expenditure)
```

> Next comes the main thing. We create a `while` loop for the user so that he can add multiple purchases and then exit whenever he wants. So first we create a loop and then if the user presses **"Y"**, we take 3 inputs.

Reviewer: phrasing
